File: 04-Arrays-Part-IV/cpp/04-longest-subarray-with-sum-k.cpp
class Solution {
public:
    int longestSubarray(vector<int>& nums, int k) {
        unordered_map<long long, int> map;

        long long sum = 0;
        int maxLen = 0;

        for (int i = 0; i < (int)nums.size(); i++) {
            sum += nums[i];

            if (sum == k) {
                maxLen = i + 1;
            }

            long long rem = sum - k;

            if (map.count(rem)) {
                maxLen = max(maxLen, i - map[rem]);
            }

            if (!map.count(sum)) {
                map[sum] = i;
            }
        }

        return maxLen;
    }
};

File: 04-Arrays-Part-IV/java/04-LongestSubarrayWithSumK.java
class Solution {
    public int longestSubarray(int[] nums, int k) {
        HashMap<Long, Integer> map = new HashMap<>();

        long sum = 0;
        int maxLen = 0;

        for (int i = 0; i < nums.length; i++) {
            sum += nums[i];

            if (sum == k) {
                maxLen = i + 1;
            }

            long rem = sum - k;

            if (map.containsKey(rem)) {
                maxLen = Math.max(maxLen, i - map.get(rem));
            }

            map.putIfAbsent(sum, i);
        }

        return maxLen;
    }
}
